### PR TITLE
Don't pan to an app if it is already on the screen

### PR DESCRIPTION
**Pull request by huwshimi**

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- add a number of apps, or deploy a bundle
- move the screen around and drag some apps around the screen and change to different zoom levels
- use the inspector to click on different charms
- the canvas should only pan to a charm if it is off screen

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7159/

File: jujugui/static/gui/src/app/views/topology/panzoom.js [hunk 1]
```diff
@@ -208,16 +208,30 @@ YUI.add('juju-topology-panzoom', function(Y) {
...
+      const pointY = (point[1] * scale) + offset[1];
+      // Find out if the points we have are within the screen rectangle.
+      const withinScreen = pointX > 0 && pointX < screenWidth && pointY > 0 &&
+        pointY < screenHeight;
+      // If the object is not within the screen the move the screen so it is
+      // visible.
+      if (!withinScreen) {
+        this.rescale({
+          scale: scale,
+          translate: point.map(function(d, i) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/views/topology/panzoom.js]: could be an arrow fn :)

```JavaScript
translate: point.map((d, i) => ((0 - d) * scale) + (size[i] / 2)))
```

Comment (makyo) [inline jujugui/static/gui/src/app/views/topology/panzoom.js]: Mixing new and old (arrowfuncs, let/const vs var) drives me up the wall.

That said, you're right 👍

**Comment by huwshimi** [comment]
Looks like this still needs a QA.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7168/

**Comment by Lukewh** [comment]
## QA-OK(ish)

This PR fixes the "Let's not move the canvas if the selected app is already fully visible" - which is so much better - but not "if it's partially visible let's just move it enough to bring it onto the canvas" or "If it's way off, then let's pan the canvas just enough to bring it fully into view". As when the charm you navigate to is "way off" it does a big'ol jumpathon.

@huwshimi can you add a couple of follow ups to implement the other two cases if they're too much of an undertaking to do as part of this PR, as I still think they're valid? Cheers

**Comment by huwshimi** [comment]
@Lukewh good call, I've left a comment in #2147. We can issue that issue to track the remaining work I think.

**Comment by huwshimi** [comment]
Thanks all :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**